File: packages/host4_flutter_aivoice/ios/AiVoice_副本/Tools/ConversationStatusMessage.m
//
//  ConversationStatusMessage.m
//  GameMacro
//
//  Created by 敬攀 on 2025/12/19.
//

#import "ConversationStatusMessage.h"

@implementation ConversationStatusMessage

+ (NSString *)unPackData:(NSData *)message{
    NSString *subtitles = convUnpack(message);
    
    return subtitles;
}

+(ConversationStatusMessage *)parseMsg:(NSString *)msg{
    return convParseData(msg);;
}

NSString *convUnpack(NSData *data) {
    const int headerSize = 8;
    
    NSUInteger size = data.length;
    
    if (size < headerSize) {
        return nil;
    }
    const uint8_t *message = data.bytes;
    // Check magic number "conv"
    uint32_t magic = (message[0] << 24) | (message[1] << 16) | (message[2] << 8) | message[3];
    
    if (magic != 0x636F6E76) {
        return nil;
    }
    
    // Get length
    uint32_t length = (message[4] << 24) | (message[5] << 16) | (message[6] << 8) | message[7];
    
    if (size - headerSize != length) {
        return nil;
    }
    
    // Get conversationStatusMessage
    NSString *conversationStatusMessage = nil;
    
    if (length > 0) {
        conversationStatusMessage = [[NSString alloc] initWithBytes:message + headerSize length:length encoding:NSUTF8StringEncoding];
    } else {
        conversationStatusMessage = @"";
    }
    
    return conversationStatusMessage;
}

ConversationStatusMessage *convParseData(NSString *message) {
    
    NSError *error = nil;
    
    NSDictionary *json_data = [NSJSONSerialization JSONObjectWithData:[message dataUsingEncoding:NSUTF8StringEncoding] options:0 error:&error];
    
    if (error || json_data == nil) {
        NSLog(@"JSON parse error: %@", error);
        return nil;
    }
    
    ConversationStatusMessage *statusMsg = [[ConversationStatusMessage alloc] init];
    
    statusMsg.taskId = json_data[@"TaskId"];
    statusMsg.userID = json_data[@"UserID"];
    statusMsg.roundID = [json_data[@"RoundID"] longLongValue];
    statusMsg.eventTime = [json_data[@"EventTime"] longLongValue];
    
    Stage stage;
    
    stage.code = [json_data[@"Stage"][@"Code"] intValue];
    stage.description = json_data[@"Stage"][@"Description"];
    statusMsg.stage = stage;
    
    if (json_data[@"ErrorInfo"] && json_data[@"ErrorInfo"] != [NSNull null]) {
        
        ErrorDetail errorDetail;
        errorDetail.code = [json_data[@"ErrorInfo"][@"ErrorCode"] intValue];
        errorDetail.reason = json_data[@"ErrorInfo"][@"Reason"];
        statusMsg.errorInfo = errorDetail;
        statusMsg.hasErrorInfo = YES;
    } else {
        statusMsg.hasErrorInfo = NO;
    }
    
    return statusMsg;
}
    
@end

File: packages/host4_flutter_aivoice/ios/Classes/Models/ConversationStatusMessage.swift
import Foundation

/// conv 消息 - 智能体状态
struct Stage {
  let code: Int
  let description: String
}

struct ErrorDetail {
  let code: Int
  let reason: String
}

/// 智能体状态消息（conv 协议）
class ConversationStatusMessage {
  let taskId: String
  let userID: String
  let roundID: Int64
  let eventTime: Int64
  let stage: Stage
  let errorInfo: ErrorDetail?
  let hasErrorInfo: Bool

  private init(taskId: String, userID: String, roundID: Int64,
               eventTime: Int64, stage: Stage,
               errorInfo: ErrorDetail?) {
    self.taskId = taskId
    self.userID = userID
    self.roundID = roundID
    self.eventTime = eventTime
    self.stage = stage
    self.errorInfo = errorInfo
    self.hasErrorInfo = errorInfo != nil
  }

  /// 拆包校验：从二进制数据中提取 conv JSON 字符串
  /// magic number = 0x636F6E76 ("conv")
  static func unpack(from data: Data) -> String? {
    let headerSize = 8
    guard data.count >= headerSize else { return nil }

    let magic = data.withUnsafeBytes { ptr in
      ptr.loadUnaligned(fromByteOffset: 0, as: UInt32.self).bigEndian
    }
    guard magic == 0x636F_6E76 else { return nil }

    let length = data.withUnsafeBytes { ptr in
      ptr.loadUnaligned(fromByteOffset: 4, as: UInt32.self).bigEndian
    }
    guard data.count - headerSize == length else { return nil }

    if length == 0 { return "" }
    return String(data: data.subdata(in: headerSize..<data.count), encoding: .utf8)
  }

  /// 解析 JSON 字符串为 ConversationStatusMessage
  static func parse(json: String) -> ConversationStatusMessage? {
    guard let jsonData = json.data(using: .utf8),
          let dict = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any]
    else { return nil }

    let stageDict = dict["Stage"] as? [String: Any]
    let stage = Stage(
      code: stageDict?["Code"] as? Int ?? 0,
      description: stageDict?["Description"] as? String ?? ""
    )

    var errorInfo: ErrorDetail? = nil
    if let errorDict = dict["ErrorInfo"] as? [String: Any],
       !(errorDict is NSNull) {
      errorInfo = ErrorDetail(
        code: errorDict["ErrorCode"] as? Int ?? 0,
        reason: errorDict["Reason"] as? String ?? ""
      )
    }

    return ConversationStatusMessage(
      taskId: dict["TaskId"] as? String ?? "",
      userID: dict["UserID"] as? String ?? "",
      roundID: (dict["RoundID"] as? Int64) ?? 0,
      eventTime: (dict["EventTime"] as? Int64) ?? 0,
      stage: stage,
      errorInfo: errorInfo
    )
  }
}
